js/app/test/cancel_events.spec.ts

```diff
@@ -0,0 +1,40 @@
+import { test, expect } from "@gradio/tootils";
+
+test("yielding", async ({ page }) => {
+	let first_iteration;
+	let fourth_iteration;
+	let last_iteration;
+
+	page.on("websocket", (ws) => {
+		first_iteration = ws.waitForEvent("framereceived", {
+			predicate: (event) => {
+				return JSON.parse(event.payload as string).msg === "process_generating";
+			}
+		});
+
+		fourth_iteration = ws.waitForEvent("framereceived", {
+			predicate: (event) => {
+				return JSON.parse(event.payload as string)?.output?.data?.[0] === "3";
+			}
+		});
+
+		last_iteration = ws.waitForEvent("framereceived", {
+			predicate: (event) => {
+				return JSON.parse(event.payload as string).msg === "process_completed";
+			}
+		});
+	});
+
+	const start_button = await page.locator("button", {
+		hasText: /Start Iterating/
+	});
+	const textbox = await page.getByLabel("Iterative Output");
+	await start_button.click();
+
+	await first_iteration;
+	await expect(textbox).toHaveValue("0");
+	await fourth_iteration;
+	await expect(textbox).toHaveValue("3");
+	await last_iteration;
+	await expect(textbox).toHaveValue("8");
```

Comment:
This is what I mean about interweaving awaits and assertions, we need to assert after we have processed a certain websocket message, then we can check that the UI correctly represents the contents of that message. The first iteration checks that we've received the first response (process_generating), and the UI correctly reflects that value.

We do the same for the others. It doesn't matter what the specifics of the predicate are as long as it waits for the correct message (roughly).